File: lib/tools/reddit.ts
import { getCached, setCache } from '../supabase';
import { searchHN } from './hn-algolia';
import type { ToolResult, RedditPost, HNPost } from './types';
import { buildToolResult } from './fallback';

// Reddit public JSON API — no key required
const BASE_URL = 'https://www.reddit.com';

function detectSentiment(text: string): 'positive' | 'negative' | 'neutral' {
  const lower = text.toLowerCase();
  const negative = ['terrible', 'awful', 'broken', 'scam', 'hate', 'disappointed', 'worst', 'bad', 'overpriced', 'useless', 'unreliable'];
  const positive = ['love', 'great', 'amazing', 'excellent', 'best', 'fantastic', 'perfect', 'awesome', 'recommend', 'worth'];
  const negScore = negative.filter(w => lower.includes(w)).length;
  const posScore = positive.filter(w => lower.includes(w)).length;
  if (negScore > posScore) return 'negative';
  if (posScore > negScore) return 'positive';
  return 'neutral';
}

// ── HN Algolia fallback — used when Reddit blocks or returns empty ─────────────
async function hnFallback(query: string): Promise<ToolResult<RedditPost[]>> {
  const hn = await searchHN(query);
  const posts: RedditPost[] = hn.data.map((p: HNPost) => ({
    title: p.title,
    subreddit: 'r/hackernews',
    score: p.score,
    url: p.url,
    snippet: p.title,
    created: p.created,
    sentiment: 'neutral' as const,
  }));
  return buildToolResult<RedditPost[]>({
    data: posts,
    status: posts.length > 0 ? 'degraded' : 'failed',
    source: 'Hacker News (Reddit fallback)',
    sourceUrl: 'https://hn.algolia.com',
  });
}

export async function searchReddit(
  query: string,
  subreddit?: string
): Promise<ToolResult<RedditPost[]>> {
  const cacheKey = `reddit:${subreddit ?? 'all'}:${query}`;
  const cached = await getCached('reddit', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<RedditPost[]>), cached: true };
  }

  const searchPath = subreddit
    ? `/r/${subreddit}/search.json`
    : '/search.json';

  const url = new URL(`${BASE_URL}${searchPath}`);
  url.searchParams.set('q', query);
  url.searchParams.set('sort', 'relevance');
  url.searchParams.set('t', 'year');
  url.searchParams.set('limit', '15');
  if (subreddit) url.searchParams.set('restrict_sr', '1');

  try {
    const res = await fetch(url.toString(), {
      headers: { 'User-Agent': 'GrowthIntelBot/1.0 (hackathon demo)' },
    });

    if (!res.ok) {
      return hnFallback(query);
    }

    const raw = await res.json() as any;
    const posts: RedditPost[] = (raw.data?.children ?? [])
      .filter((c: any) => c.kind === 't3')
      .slice(0, 10)
      .map((c: any) => {
        const p = c.data;
        const text = `${p.title} ${p.selftext ?? ''}`;
        return {
          title: p.title,
          subreddit: p.subreddit_name_prefixed,
          score: p.score,
          url: `https://reddit.com${p.permalink}`,
          snippet: p.selftext ? p.selftext.slice(0, 300) : p.title,
          created: new Date(p.created_utc * 1000).toISOString(),
          sentiment: detectSentiment(text),
        };
      });

    // Reddit returned nothing — fall back to HN
    if (posts.length === 0) return hnFallback(query);

    const result = buildToolResult<RedditPost[]>({
      data: posts,
      status: 'ok',
      source: 'Reddit',
      sourceUrl: url.toString(),
    });

    await setCache('reddit', cacheKey, result);
    return result;
  } catch {
    return hnFallback(query);
  }
}

export async function searchProductReviews(productName: string): Promise<ToolResult<RedditPost[]>> {
  return searchReddit(`${productName} review OR experience OR pricing OR alternative`);
}

export async function searchSubreddits(
  query: string,
  subreddits: string[]
): Promise<ToolResult<RedditPost[]>> {
  const results = await Promise.allSettled(
    subreddits.map(sr => searchReddit(query, sr))
  );

  const rejectedCount = results.filter(r => r.status === 'rejected').length;

  const allPosts: RedditPost[] = [];
  for (const r of results) {
    if (r.status === 'fulfilled') {
      allPosts.push(...r.value.data);
    }
  }

  // Deduplicate and sort by score
  const seen = new Set<string>();
  const unique = allPosts
    .filter(p => {
      if (seen.has(p.url)) return false;
      seen.add(p.url);
      return true;
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, 12);

  const status = unique.length === 0
    ? 'failed'
    : rejectedCount > 0
      ? 'degraded'
      : 'ok';

  return buildToolResult<RedditPost[]>({
    data: unique,
    status,
    source: 'Reddit (multi-subreddit)',
    sourceUrl: 'https://www.reddit.com',
  });
}

File: backend/app/tools/reddit.py
"""Reddit public JSON API — mirrors lib/tools/reddit.ts."""

from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urlencode

from app.supabase_client import get_cached, set_cache

from .fallback import build_tool_result
from .hn_algolia import search_hn

BASE_URL = "https://www.reddit.com"
USER_AGENT = "GrowthIntelBot/1.0 (hackathon demo)"
TIMEOUT = 15.0

_NEGATIVE = [
    "terrible", "awful", "broken", "scam", "hate", "disappointed",
    "worst", "bad", "overpriced", "useless", "unreliable",
]
_POSITIVE = [
    "love", "great", "amazing", "excellent", "best", "fantastic",
    "perfect", "awesome", "recommend", "worth",
]


def _detect_sentiment(text: str) -> str:
    lower = text.lower()
    neg_score = sum(1 for w in _NEGATIVE if w in lower)
    pos_score = sum(1 for w in _POSITIVE if w in lower)
    if neg_score > pos_score:
        return "negative"
    if pos_score > neg_score:
        return "positive"
    return "neutral"


async def _hn_fallback(query: str) -> dict[str, Any]:
    hn = await search_hn(query)
    posts = [
        {
            "title": p.get("title", ""),
            "subreddit": "r/hackernews",
            "score": p.get("score", 0),
            "url": p.get("url", ""),
            "snippet": p.get("title", ""),
            "created": p.get("created", ""),
            "sentiment": "neutral",
        }
        for p in (hn.get("data") or [])
    ]
    return build_tool_result(
        data=posts,
        status="degraded" if posts else "failed",
        source="Hacker News (Reddit fallback)",
        source_url="https://hn.algolia.com",
    )


async def search_reddit(query: str, subreddit: str | None = None) -> dict[str, Any]:
    cache_key = f"reddit:{subreddit or 'all'}:{query}"
    cached = await get_cached("reddit", cache_key)
    if cached:
        return {**cached, "cached": True}

    search_path = f"/r/{subreddit}/search.json" if subreddit else "/search.json"
    params: dict[str, str] = {
        "q": query,
        "sort": "relevance",
        "t": "year",
        "limit": "15",
    }
    if subreddit:
        params["restrict_sr"] = "1"

    url = f"{BASE_URL}{search_path}?{urlencode(params)}"

    try:
        import httpx

        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            res = await client.get(url, headers={"User-Agent": USER_AGENT})

        if not res.is_success:
            return await _hn_fallback(query)

        raw = res.json()
        posts: list[dict[str, Any]] = []
        for c in (raw.get("data", {}).get("children") or []):
            if c.get("kind") != "t3":
                continue
            p = c.get("data") or {}
            text = f"{p.get('title', '')} {p.get('selftext') or ''}"
            posts.append(
                {
                    "title": p.get("title", ""),
                    "subreddit": p.get("subreddit_name_prefixed", ""),
                    "score": p.get("score", 0),
                    "url": f"https://reddit.com{p.get('permalink', '')}",
                    "snippet": (p.get("selftext") or p.get("title", ""))[:300],
                    "created": datetime.fromtimestamp(
                        p.get("created_utc", 0), tz=timezone.utc
                    ).isoformat(),
                    "sentiment": _detect_sentiment(text),
                }
            )
            if len(posts) >= 10:
                break

        if not posts:
            return await _hn_fallback(query)

        result = build_tool_result(
            data=posts,
            status="ok",
            source="Reddit",
            source_url=url,
        )
        await set_cache("reddit", cache_key, result)
        return result
    except Exception:
        return await _hn_fallback(query)


async def search_product_reviews(product_name: str) -> dict[str, Any]:
    return await search_reddit(
        f"{product_name} review OR experience OR pricing OR alternative"
    )


async def search_subreddits(query: str, subreddits: list[str]) -> dict[str, Any]:
    results = await asyncio.gather(
        *[search_reddit(query, sr) for sr in subreddits],
        return_exceptions=True,
    )

    rejected_count = sum(1 for r in results if isinstance(r, BaseException))
    all_posts: list[dict[str, Any]] = []
    for r in results:
        if isinstance(r, dict):
            all_posts.extend(r.get("data") or [])

    seen: set[str] = set()
    unique: list[dict[str, Any]] = []
    for p in sorted(all_posts, key=lambda x: x.get("score", 0), reverse=True):
        post_url = p.get("url", "")
        if post_url in seen:
            continue
        seen.add(post_url)
        unique.append(p)
        if len(unique) >= 12:
            break

    if not unique:
        status = "failed"
    elif rejected_count > 0:
        status = "degraded"
    else:
        status = "ok"

    return build_tool_result(
        data=unique,
        status=status,
        source="Reddit (multi-subreddit)",
        source_url="https://www.reddit.com",
    )
